proposal: Falco API (#908)
falcosecurity/falco: proposals/20191030-api.md

# Falco APIs

## Summary

> This is a proposal to better structure the Falco API.

Reviewer: I think we should spend a few minutes talking about what exactly the Falco API is. As well as highlighting the problem(s) that it would potentially be solving.

I know we go into detail later, but a sentence or two here might be useful.

> It will contain a list of gRPC services that can be added to have a first stable version of the API for v1.0.0.

Reviewer: So is gRPC the API? or is gRPC the first implementation of a way of exposing the API over a network socket?

I think maybe the API is simply the agreement of the shape of objects, and what they represent and do right?

## Motivation

We want to enable users to use thirdy-party clients to interface with Falco outputs, inputs, rules, and configurations.

Such ability would enable the community to create a whole set of OSS tools, built on top of Falco.

Some examples, already in place, are:

- [falcoctl](https://github.com/falcosecurity/falcoctl)
- [falco-exporter](https://github.com/falcosecurity/falco-exporter)

## Goals

- To design and implement an API to make Falco more configurable
- To design and implement an API to make Falco more consumable
- To design a set of simple contract interfaces
- To have a secure by default Falco API

## Non-Goals

- Replace the existing inputs: the **inputs** will be only for new source events users may want to add, initially

> 

Reviewer: What about documentation? Is that a non-goal here? If so we should maybe call it out?

## Proposal

### Use cases

> - Receive Falco events with a well-defined contract over wire --> **outputs**

Reviewer: remove wire? again this goes back to my previous comment about the API being a shape and agreement and not an implementation.

For instance if a user wanted to vendor falco in C++ they would still potentially be using "an API"

- Receive Falco drops with a well-defined contract over wire --> **drops**
- Receive current Falco version and related meta information (commit hash, built type, etc.) --> **version**
- Configure Falco via API (CRUD) -> **configs**

> - Inject and/or modify Falco rules via API (CRUD) --> **rules**

Reviewer 2: This would be awesome but would still require a route into the kubernetes cluster (or instance) to configure these or a sidecar that is used for configuration.

I am fairly new to the Falco project but has any thought been given to a central management server for Falco agents to report into and fetch new configurations / rules dynamically rather than having to hit an API or reload rules manually? I know this would be a new project to be spun up of the management server but interested if this has been proposed / discussed in the past before

Author: Maybe something like [this](https://github.com/falcosecurity/falco/issues/824) can help with what are you suggesting?

Reviewer: WRT having any external configuration management, the work starts here with implementing a clean interface around the core of the program. 

I think we would all be very excited to see work like a sidecar or configuration management server come to life, and the Falco community would be happy to sponsor this work. 

If you have ideas for configuration management we could open up a new proposal (like this one) with your ideas, and get a repository and a Falco client set up so work can begin building out something like what you described :smile:

# Falco APIs

## Summary

> With this proposal we inted to present the intent of creating a Falco API.

Reviewer 3: ```suggestion
This is a proposal to better structure the Falco API.
```

> This proposal also intends to list the set of the gRPC services such API will expose.

Reviewer 3: ```suggestion
It will contain a list of gRPC services that can be added to have a first stable version of the API for v1.0.0.
```

## Motivation

> We want users to interface with Falco outputs, inputs, rules, and configurations via thirdy-part clients.

Reviewer 3: ```suggestion
We want to enable users to use thirdy-party clients to interface with Falco outputs, inputs, rules, and configurations.
```

Such ability would enable the community to create a whole set of OSS tools, built on top of Falco.

Some examples, already in place, are:

- [falcoctl](https://github.com/falcosecurity/falcoctl)
- [falco-exporter](https://github.com/falcosecurity/falco-exporter)

## Goals

- To design and implement an API to make Falco more configurable
- To design and implement an API to make Falco more consumable
- To design a set of simple contract interfaces
- To have a secure by default Falco API

## Non-Goals

-

## Proposal

### Use cases

- Receive Falco events with a well-defined contract over wire --> **outputs**
- Receive Falco drops with a well-defined contract over wire --> **drops**
- Receive current Falco version and related meta information (commit hash, built type, etc.) --> **version**

> - Config Falco via API (CRUD) -> **configs**

Reviewer 3: ```suggestion
- Configure Falco via API (CRUD) -> **configs**
```

    message request
    {
      ...
    }

    // The `response` message contains the version of Falco.
    // It provides the whole version as a string and also
    // its parts as per semver 2.0 specification (https://semver.org).
    message response
    {
    string version = 1;
      uint32 major = 2;
      uint32 minor = 3;
      uint32 patch = 4;
      string prerelease = 5;
      string build = 6;
    }
    ```

- one or more `.proto` containing the commond models - ie., the already existing `schema.proto` containing source enum, etc.

    ```protobuf
    # schema.proto
    enum priority {
      ..
    }
    enum source {
    option allow_alias = true;
      SYSCALL = 0;
      syscall = 0;
      Syscall = 0;
      K8S_AUDIT = 1;
      k8s_audit = 1;
      K8s_audit = 1;
      K8S_audit = 1;
    }
    ```

- a `.proto` for service definitions

    For example
    ```protobuf
    service unary {
      rpc version(request) returns (response);
      ...
    }
    service streaming {
      rpc outputs(request) returns (stream response);
      rpc drops(request) returns (stream response);
    }
    service bidirectional {
      ...
    }
    ```

#### Benefits

> We think this structure makes the implementation more easy to write and to maintain because it automatically matches to abstractions - ie., one service by RPC type, a pool of contexts by RPC type - implemented in the code.

Reviewer 3: ```suggestion
We think this structure makes the implementation easier to write and to maintain because it automatically matches the abstractions - ie., one service by RPC type, a pool of contexts by RPC type - implemented in the code.
```
